main.go

```diff
@@ -18,11 +20,196 @@ import (
 	"github.com/nats-io/nats.go"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/viper"
+	"go.opentelemetry.io/otel"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
 var log = logging.Log.WithFields(logrus.Fields{"service": "discoenv-users-service"})
 
+// NATSMessageHandler is the handler type that we support. Should map to the nats.Handler
+// function signature func(*nats.Msg).
+type NATSMessageHandler func(context.Context, *nats.Msg)
+
+type NATSTextMapCarrier struct {
+	nats.Header
+}
+
+func (n NATSTextMapCarrier) Keys() []string {
+	var keys []string
+	for key := range n.Header {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+func ProtoUnmarshal(data []byte, ptr interface{}) error {
+	if _, ok := ptr.(*interface{}); ok {
+		return nil
+	}
+
+	msg, ok := ptr.(proto.Message)
+	if !ok {
+		return errors.New("invalid protocol buffer message passed to ProtoUnmarshal")
+	}
+
+	return protojson.Unmarshal(data, msg)
+}
+
+// func ProtoMarshal(ptr interface{}) ([]byte, error) {
+// 	msg, ok := ptr.(proto.Message)
+// 	if !ok {
+// 		return nil, errors.New("invalid protocol buffer message passed to ProtoMarshal")
+// 	}
+// 	b, err := protojson.Marshal(msg)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	return b, nil
+// }
+
+type NATSOtelEncodedConnection struct {
+	nats.EncodedConn
+	OtelInstrumentationName string
+	OtelSystem              string
+	OtelProtocol            string
+	OtelProtocolVersion     string
+}
+
+// Middleware adds trace info to the message headers. Only accepts
+// nats.Handler of the form func(*nats.Msg) for now. Anything else will
+// bypass the telemetry logic.
+func (ec *NATSOtelEncodedConnection) Middleware(ctx context.Context, next NATSMessageHandler) NATSMessageHandler {
+	return func(ctx context.Context, msg *nats.Msg) {
+		headerCarrier := NATSTextMapCarrier{
+			msg.Header,
+		}
+		msgCtx := otel.GetTextMapPropagator().Extract(ctx, headerCarrier)
+		tracer := otel.GetTracerProvider().Tracer(ec.OtelInstrumentationName)
+		msgCtx, span := tracer.Start(msgCtx, msg.Subject+" process", trace.WithSpanKind(trace.SpanKindConsumer))
+		defer span.End()
+
+		span.SetAttributes(
+			semconv.MessagingSystemKey.String(ec.OtelSystem),
+			semconv.MessagingProtocolKey.String(ec.OtelProtocol),
+			semconv.MessagingProtocolVersionKey.String(ec.OtelProtocolVersion),
+			semconv.MessagingDestinationKindTopic.Key.String(msg.Subject),
+			semconv.MessagingOperationKey.String("process"),
+		)
+		next(msgCtx, msg)
+	}
+}
+
+func NewNATSOtelEncodedConnection(encConn *nats.EncodedConn, instr, system, proto, protoVersion string) *NATSOtelEncodedConnection {
```

Comment:
Not sure exactly where to add this comment so it's just here but:

it feels like this is... I dunno, doing too much, I guess? It seems like the opentelemetry stuff should be able to be a very thin wrapper in general -- probably one thing that takes in a `func (context.Context, <whatever else>)` and returns an appropriate handler (presumably, a `func (<whatever else>)`). And then you'd just do the usual `QueueSubscribe` but with `QueueSubscribe(*subject, *queue, ThisWrapperThing(theActualHandler))` instead of `QueueSubscribe(*subject, *queue, theActualHandler)`. Similarly on the publish side, I'd expect something that allows passing in a context and which basically does nothing but create a message as usual, inject the context, and then pass it down to the normal nats code.

I guess maybe this is because of how the EncodedConn works, since it seems to not have... anything for using headers along with encoded connections. Which seems like an oversight on their part, to say the least, lol.

Reply:
Yeah, this is fully because the only way to get through to headers in an EncodedConnection seems to be with the `func(*nats.Msg)` handler signature for nats.Handler (NATS uses reflection to determine how to call their handlers). Plus, after digging in, I didn't see/think of a better way to instrument the `Publish*()` functions.

I was mildly annoyed by the amount of code I was having to write, too, so I was thinking that maybe putting the fields we need into the protocol buffers themselves might actually be a better implementation, overall. Then we could have a custom encoder that handles the JSON encoding/decoding and can also manage extracting/embedding telemetry info rather than having to write a bunch of code to handle it before delegating to the EncodedConnection.

Reply:
This should be fixed now.